### PR TITLE
Add a 'public' property to control field visibility

### PR DESCRIPTION
**Pull request by alafon**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? |no
| Tests pass?   | yes
| Documented?   | not yet
| Fixed tickets | n/a
| License       | MIT

Make it possible to completely remove a field (from any queries, including inspection queries) based on the Expression Language, by not exposing it.

When defining fields, `public` works the same way as `access`, except that defining a default `public` behavior is not supported because it does not make sens.

### CODE REVIEW

File: Resources/skeleton/TypeSystem.php.skeleton [hunk 1]
```diff
@@ -27,4 +27,19 @@
 <spaces><spaces>$class = substr(strrchr(__CLASS__, "\\"), 1);
 <spaces><spaces>return substr($class, 0, -4);
 <spaces>}
+
+<spaces>private static function applyExposeFilters($fields)
```
Comment (mcg-web) [inline Resources/skeleton/TypeSystem.php.skeleton]: Why this is static function?

File: Generator/TypeGenerator.php [hunk 2]
```diff
@@ -72,6 +82,34 @@ protected function resolveTypeCode($alias)
         return  sprintf('$container->get(\'%s\')->resolve(%s)', 'overblog_graphql.type_resolver', var_export($alias, true));
     }
 
+    protected function generateExpose(array $value)
+    {
+        $exposeIsSet = $this->arrayKeyExistsAndIsNotNull($value, 'expose');
+        $fieldOptions = $value;
+        if (!$this->arrayKeyExistsAndIsNotNull($fieldOptions, 'expose')) {
+            $fieldOptions['expose'] = function () {
```
Comment (alafon) [inline Generator/TypeGenerator.php]: `return 'null'` instead of the callback function

File: Generator/TypeGenerator.php [hunk 3]
```diff
@@ -62,6 +62,16 @@ protected function generateClassDocBlock(array $value)
 EOF;
     }
 
+    protected function generateOutputFields(array $config)
+    {
+        $outputFieldsCode = sprintf(
+            '$this->applyExposeFilters(%s)',
```
Comment (alafon) [inline Generator/TypeGenerator.php]: Use `self::` instead of `$this`

**Comment by alafon** [comment]
- comments from @mcg-web taken into account
- commits rebased
- doc added

With ❤️

**Comment by ooflorent** [comment]
I don't like the naming. Could you rename the field `public` to match Symfony's DI config?

**Comment by ooflorent** [comment]
Another issue is schema generation. If you want to use Relay, use MUST generate the schema upfront. Private fields will then be an issue...

**Comment by alafon** [comment]
@ooflorent do you think that it's a blocker for the PR to be merged? If you think of a specific use case where it would lead to unwanted situation please let me know.

**Comment by mcg-web** [comment]
No this is not a blocker to me... can you squash your commits please? then i'll merge this.

**Comment by alafon** [comment]
Done, thanks 👍

**Pull request merged**